geopyspark/geotrellis/tms.py

```diff
@@ -38,83 +46,138 @@ def render(self, cells, cols, rows): # return `bytes`
-            catalog (string): The name of the catalog
-            color_map (ColorMap): A ColorMap to use in rendering the catalog tiles
+            render_function (list[numpy array] => bytes): A function to convert
+                a list of numpy arrays to a collection of bytes giving a binary
+                image file.
 
         Returns:
-            :class:`~geopyspark.geotrellis.tms.TMSServer`
+            [TileCompositer]
         """
+        # composite_function: List[np.array] => Image
+        self.composite_function = composite_function
 
-        server = pysc._gateway.jvm.geopyspark.geotrellis.tms.TMSServer.serveS3Catalog(bucket, root, catalog, color_map.cmap)
-        return cls(pysc, server)
+    def requiresEncoding(self):
+        return True
 
-    @classmethod
-    def remote_tms_server(cls, pysc, pattern_url):
-        """A function to create a TMS server delivering tiles from a remote TMS server
+    def compositeEncoded(self, all_scala_arrays): # return `bytes`
+        """A function to convert an array to an image.
 
         Args:
-            pattern_url (string): A string giving the form of the URL where tiles
-                are stored.  The pattern should contain the literals '{z}', '{x}',
-                and '{y}' giving the zoom, x, and y keys of the desired tile,
-                respectively.
+            all_scala_arrays (array of bytes): An array containing the encoded
+                representations of the incoming tiles
 
         Returns:
-            :class:`~geopyspark.geotrellis.tms.TMSServer`
+            [bytes] representing an image
         """
+        try:
+            cells = [multibandtile_decoder(scala_array).cells for scala_array in all_scala_arrays]
+            image = self.composite_function(cells)
+            bio = io.BytesIO()
+            image.save(bio, 'PNG')
+            return bio.getvalue()
+        except Exception:
+            from traceback import print_exc
+            print_exc()
 
-        server = pysc._gateway.jvm.geopyspark.geotrellis.tms.TMSServer.serveRemoteTMSLayer(pattern_url)
-        return cls(pysc, server)
+    class Java:
+        implements = ["geopyspark.geotrellis.tms.TileCompositer"]
+
+class TMS(object):
+    """Provides a TMS server for raster data.
+
+    In order to display raster data on a variety of different map interfaces
+    (e.g., leaflet maps, geojson.io, GeoNotebook, and others), we provide
+    the TMS class.
+    """
+    def __init__(self, pysc, server):
+        self.pysc = pysc
+        self.server = server
+        self.handshake = ''
+        pysc._gateway.start_callback_server()
+
+    def set_handshake(self, handshake):
+        self.server.set_handshake(handshake)
+        self.handshake = handshake
 
     @classmethod
-    def rdd_tms_server(cls, pysc, pyramid, color_map):
-        """Creates a ``TMSServer`` from a ``Pyramid`` instance.
+    def build(cls, pysc, *args, **kwargs):
```

Comment:
Would something like this work for the `*args`?

```python
def build(cls, pysc, source, *args, renderer=None, compositor=None)
```

That way the user knows that there needs to be at least one `source` parameter with the option of passing in more to the method. I'm also still not sure why the `renderer` and the `compositor` can't just be named parameters with a default of `None` for both. With the distinction between the two being made in the documentation. The user already has to specify which one they're using, this just makes it easier to discover the argument names.